utt/signature-processor/src/sigProcessor.cpp

```diff
@@ -0,0 +1,181 @@
+// file.
+
+#include "sigProcessor.hpp"
+#include "mint.hpp"
+#include "common.hpp"
+#include "coinsSigner.hpp"
+#include "budget.hpp"
+#include "transaction.hpp"
+#include <utt/Comm.h>
+
+#include "bftengine/ReplicaConfig.hpp"
+#include "bftengine/messages/MessageBase.hpp"
+#include "bftengine/messages/ClientRequestMsg.hpp"
+
+namespace utt {
+namespace messages {
+class PartialSigMsg : public bftEngine::impl::MessageBase {
+ public:
+  PartialSigMsg(uint16_t srcReplicaId, const libutt::api::types::Signature& partialSig, uint64_t sigId)
+      : bftEngine::impl::MessageBase(
+            srcReplicaId, bftEngine::impl::MsgCode::Reserved, 0, sizeof(Header) + (uint32_t)partialSig.size()) {
+    b()->partialSigLength = (uint16_t)partialSig.size();
+    b()->sigId = sigId;
+    b()->genReplicaId = srcReplicaId;
+    char* position = body() + sizeof(Header);
+    memcpy(position, partialSig.data(), partialSig.size());
+  }
+
+  BFTENGINE_GEN_CONSTRUCT_FROM_BASE_MESSAGE(PartialSigMsg)
+
+  libutt::api::types::Signature getPartialSig() const {
+    char* position = body() + sizeof(Header);
+    libutt::api::types::Signature ret(b()->partialSigLength);
+    memcpy(ret.data(), position, b()->partialSigLength);
+    return ret;
+  }
+  uint16_t idOfGeneratedReplica() const { return b()->genReplicaId; }
+
+  uint16_t partialSignatureLen() const { return b()->partialSigLength; }
+
+  uint64_t getSigId() const { return b()->sigId; }
+
+  void validate(const ReplicasInfo&) const override {}
+
+  static uint32_t sizeOfHeader() { return sizeof(Header); }
+
+ private:
+#pragma pack(push, 1)
+
+  struct Header : public bftEngine::impl::MessageBase::Header {
+    uint16_t genReplicaId;
+    uint64_t sigId;
+    uint16_t partialSigLength;
+  };
+#pragma pack(pop)
+  static_assert(sizeof(Header) == (6 + 2 + 8 + 2), "Header is 18B");
+
+  Header* b() const { return (Header*)msgBody_; }
+};
+}  // namespace messages
+SigProcessor::SigProcessor(uint16_t repId,
+                           uint16_t n,
+                           uint16_t threshold,
+                           uint64_t timer_handler_timeout,
+                           std::shared_ptr<bftEngine::impl::MsgsCommunicator> msgsCommunicator,
+                           std::shared_ptr<bftEngine::impl::MsgHandlersRegistrator> msgHandlers,
+                           concordUtil::Timers& timers)
+    : msgsCommunicator_{msgsCommunicator},
+      timers_{timers},
+      repId_{repId},
+      n_{n},
+      threshold_{threshold},
+      timer_handler_timeout_{timer_handler_timeout} {
+  msgHandlers->registerMsgHandler(bftEngine::impl::MsgCode::Reserved, [&](bftEngine::impl::MessageBase* message) {
+    messages::PartialSigMsg* msg = (messages::PartialSigMsg*)message;
+    onReceivingNewPartialSig(msg->getSigId(), msg->idOfGeneratedReplica(), msg->getPartialSig());
+    delete msg;
+  });
+  timeout_handler_ = timers_.add(std::chrono::milliseconds(timer_handler_timeout_),
```

Comment:
I see that in the test, the timer evaluation is called from ```IncomingMsgsStorageImp::dispatchMessages``` , is it going to be the same as in the product? if the job entry is big, can it block the msg dispatcher for a significant time?

Reply:
yes, this is the stright forward way we have to implement a thread-safe timer.
You are right that if we have many ongoing signature processing this may block the message thread, but I don't think this is will be a frequent scenario. Recall that once we have done computing the signatures we evacuate the relevant entry.
Anyway, I think we can leave it as it is for now, and lets keep it in mind for the futuer